src/test/java/com/shapesecurity/salvation2/ParserTest.java

```diff
@@ -224,6 +219,21 @@ public void testPreservesMalformed() {
 				"upgrade-insecure-requests a",
 				e(Policy.Severity.Error, "The upgrade-insecure-requests directive does not support values", 0, 0)
 		);
+
+		roundTrips(
+			"&",
+			e(Policy.Severity.Error, "Directive name & contains characters outside the range  ALPHA / DIGIT / \"-\"", 0, -1)
+		);
+
+		roundTrips(
+			"default-src'self'",
+			e(Policy.Severity.Error, "Directive name default-src'self' contains characters outside the range  ALPHA / DIGIT / \"-\"", 0, -1)
+		);
+
+		roundTrips(
+			"default-src'self' a",
```

Comment:
This should apply to any directive right? So maybe add a test where it’s the third or fourth directive that’s malformed?

Reply:
Is lack of space in values handled? Ex if this was `’self’a`

Sorry if I’m straying too far from the original issue.

Reply:
> This should apply to any directive right? So maybe add a test where it’s the third or fourth directive that’s malformed?

It should, yes. Right now we're not really doing those sorts of tests; there's nothing special about any particular position. If we want to get comprehensive coverage for the combinatorial explosion of possible ways of composing things we'll need a fuzzer, I expect; I don't think it's worth doing piecemeal.

> Is lack of space in values handled?

`default-src'self'a` is treated as a single (malformed) directive name with no values (which matches browsers). `default-src 'self'a` is treated as a `default-src` directive with an unknown source-expression (which also matches browsers). I've added both as tests.

Reply:
Fair, I didn’t have much of a look at the code flow as far as whether or not positionality mattered. I just figured it was worth being clear that it wasn’t limited to default-src. I definitely agree we don’t need to get into fuzzing.

Thanks for adding tests for the others !